std/array.d

```diff
@@ -192,13 +192,21 @@ if (isIterable!Range && !isNarrowString!Range && !isInfinite!Range)
 
 /**
 Convert a narrow string to an array type that fully supports random access.
-This is handled as a special case and always returns a $(D dchar[]),
-$(D const(dchar)[]), or $(D immutable(dchar)[]) depending on the constness of
-the input.
+This is handled as a special case and always returns an array of `dchar`
+
+Params:
+    str = `isNarrowString` to be converted to an array of `dchar`
+Returns:
+    a $(D dchar[]), $(D const(dchar)[]), or $(D immutable(dchar)[]) depending on the constness of
+    the input.
 */
-ElementType!String[] array(String)(String str) if (isNarrowString!String)
+@trusted ElementType!String[] array(String)(scope String str) if (isNarrowString!String)
 {
     import std.utf : toUTF32;
+    /* This function is @trusted because the following cast
+     * is unsafe - converting a dstring[] to a dchar[], for example.
+     * Our special knowledge of toUTF32 is needed to know the cast is safe.
+     */
     return cast(typeof(return)) str.toUTF32;
```

Comment:
The whole trusted thing can be avoided by

1. Changing this line to `return toUTFImpl!(dchar[])(str);`
2. Changing `std.utf.toUTFImpl` to be `package` rather than `private`